File: Petrov/Pirson.go
package main

import (
	"math"
)

type Pirson struct {
	count map[int32]int32
	l     int32
	y     map[int32]float64
}

type PirsonAgrs<T> struct {
	x
}

func NewPirson(args PirsonAgrs) {

}

func (this Pirson) get_stat() float64 {
	n := this.count
	l := float64(this.l)
	p := this.y

	var X float64
	for i, pi := range p {
		X += math.Pow(float64(float64(n[i])-l*pi), 2.0) / (l * pi)
	}
	return X
}

func (this *Pirson) counter(arr []int32) map[int32]int32 {
	out := make(map[int32]int32, len(arr))
	for i := range arr {
		out[int32(i)]++
	}
	return out
}

func (this *Pirson) transform_range_to_array() {

}

File: Petrov/criterion/pirson.py
import abc

class Pirson(object):
    
    def __init__(self, x, y, r=False, interval=None, c=None, low=6, disable_check_len_interval=False):
        self.rnd = random.Random()
        if not r:
            self.count = Counter(x)
        elif isinstance(interval, list):
            self.count = Pirson.transform_range_to_array(x, set_range=interval, disable_check_len_interval=disable_check_len_interval)
        else:
            z = x if len(x)<=len(y) else y
            self.range = Pirson.find_optimal_interval(z, c or len(z)//2, low)
            self.count = self.range.y
        if isinstance(y, dict):
            self.y = y
            self.l = len(x)
        else:
            self.l = len(y)
            self.y = Pirson.get_squense_from_array(y, interval=self.range)

    

    @abc.abstractmethod
    def get_stat(self):
        pass